kernel/src/hil/i2c.rs

```diff
@@ -230,6 +230,7 @@ pub trait I2CDevice {
     fn write(&self, data: &'static mut [u8], len: usize) -> Result<(), (Error, &'static mut [u8])>;
     fn read(&self, buffer: &'static mut [u8], len: usize)
         -> Result<(), (Error, &'static mut [u8])>;
+    fn set_address(&self, addr: u8);
```

Comment:
We can't add this API. The address is fixed, and grants the holder of an `I2CDevice` access to only the specified device (address).

> /// It gives an interface for communicating with a specific I2C
> /// device.

Reply:
What is the device address changes though? It's not uncommon to have programmable I2C addresses

Reply:
I think either a capsule could use the `I2CMaster` interface, or a shim layer that tracks the address changes below the `I2CDevice` would work.

Reply:
I agree with @bradjc, for this abstraction, anyway, the whole point of the `I2CDevice` trait is that it isolates drivers talking to different peripherals over the same I2C bus. If a client of `I2CDevice` can modify the address arbitrarily, it completely negates the point of this abstraction.

There _could_ be a _different_ interface. For example, an extension trait might work, but I'd really want to understand the use case. Is the idea that some application/driver would change the I2C device dynamically after boot? What's a specific case for this?

Reply:
The goal is to be able to share an I2C bus with multiple kernel drivers and with userspace. Something like:

```
                             App1    App2   
                            =====syscall=====
Sensor         Screen        I2CMasterDriver
---------     ---------     -----------------
I2CDevice     I2CDevice         I2CDevice(?)
---------------------------------------------
                 Virtual I2C
---------------------------------------------
                   I2C bus
```